plugins/formal-verification/testData/diagnostics/verifies/user_invariants/simple.fir.diag.txt

```diff
@@ -0,0 +1,121 @@
+/simple.kt:(239,243): info: Generated Viper text for test:
+field bf$size: Ref
+
+method f$test$TF$T$Int(p$n: Ref) returns (ret$0: Ref)
+  ensures df$rt$isSubtype(df$rt$typeOf(ret$0), df$rt$unitType())
+{
+  var l0$it: Ref
+  var l0$holds: Ref
+  var anon$0: Ref
+  inhale df$rt$isSubtype(df$rt$typeOf(p$n), df$rt$intType())
+  l0$it := df$rt$intToRef(0)
+  l0$holds := df$rt$boolToRef(true)
+  label lbl$continue$0
+    invariant df$rt$isSubtype(df$rt$typeOf(l0$it), df$rt$intType())
+    invariant df$rt$isSubtype(df$rt$typeOf(l0$holds), df$rt$boolType())
+    invariant df$rt$isSubtype(df$rt$typeOf(p$n), df$rt$intType())
+    invariant df$rt$intFromRef(l0$it) <= 10
+    invariant df$rt$boolFromRef(l0$holds)
+  anon$0 := sp$ltInts(l0$it, df$rt$intToRef(10))
+  if (df$rt$boolFromRef(anon$0)) {
+    l0$it := sp$plusInts(l0$it, df$rt$intToRef(1))
+    goto lbl$continue$0
+  }
+  label lbl$break$0
+    invariant df$rt$isSubtype(df$rt$typeOf(l0$it), df$rt$intType())
+    invariant df$rt$isSubtype(df$rt$typeOf(l0$holds), df$rt$boolType())
+    invariant df$rt$isSubtype(df$rt$typeOf(p$n), df$rt$intType())
+    invariant df$rt$intFromRef(l0$it) <= 10
+    invariant df$rt$boolFromRef(l0$holds)
+  assert df$rt$intFromRef(l0$it) == 10
+  if (df$rt$intFromRef(l0$it) <= df$rt$intFromRef(p$n)) {
+    var anon$1: Ref
+    label lbl$continue$1
+      invariant df$rt$isSubtype(df$rt$typeOf(l0$it), df$rt$intType())
+      invariant df$rt$isSubtype(df$rt$typeOf(l0$holds), df$rt$boolType())
+      invariant df$rt$isSubtype(df$rt$typeOf(p$n), df$rt$intType())
+      invariant df$rt$intFromRef(l0$it) <= df$rt$intFromRef(p$n)
+      invariant df$rt$boolFromRef(l0$holds)
+    anon$1 := sp$ltInts(l0$it, p$n)
+    if (df$rt$boolFromRef(anon$1)) {
+      l0$it := sp$plusInts(l0$it, df$rt$intToRef(1))
+      goto lbl$continue$1
+    }
+    label lbl$break$1
```

Comment:
The fact we don't get the negation of the condition here makes the system quite a bit less powerful.  Could you make sure we've noted down this limitation somewhere?

Reply:
Hm, it's rather strange. If Viper does not assume a negation of the loop conditions here (which is reasonable) how does it prove that after a loop `it == n` here (or `it == 10` before)?

Reply:
(And even passes the test with break.)

Reply:
That's an interesting question.  Could you look into it?  A smaller example may be worth considering.  Maybe it's aware that no break can occur?